### PR TITLE
New i18n logic (arb)

### PR DESCRIPTION
**Pull request by jozsefsallai**
This PR implements a new i18n strategy which uses arb files instead of json files, for extended functionality. It also implements a way to categorize/organize the language strings so that they can be easily readable/trackable, and will look less confusing on crowdin.

**Note that this PR adds Node as a dependency of the project, however, no external packages are necessary.**

Unfortunately, the only downfall of this is the need of using static getters in the `localization.dart` service. Since getters are regular variables, the only way to semantically organize the translation strings is using underscores. Previously, the JSON method used `ALL_CAPS.AND.DOTS` to organize the strings, now these are `all_lowercase__and_separated__by_double_underscore`. A linter ignore rule has been added to the localization service file to avoid warnings about inconsistent naming conventions.

## Small "How To" Section

If, at any point in time, we want to add a new localization string, we have to do it in `lib/services/localization.dart`, as a new getter. This file will be quite big after a while, but these are the limitations of the arb approach. In the future, I'll try to somehow organize the getters so that they don't take up so much space in the file.

Once you add the localization strings, you have to run the `make generate_locale` command. This will generate the arb file with all the localization strings, and then organize them using a Node.js script. the contents of the `en` folder will then have to be uploaded to Crowdin where the system will automatically do everything necessary for the other languages. Once the translations are added on Crowdin, build the project and download the translations.

Then copy the language files to the `assets/i18n` folder and run `make build_locale`. which will unify the arb files into one arb file, for each locale respectively, and then generate the necessary Dart files to be used in `lib/locale`.

### TL;DR:

  - `make generate_locale`
  - upload `assets/i18n/en` folder's contents to crowdin
  - apply the translations on crowdin, then build and download
  - copy the contents of the downloaded archive to `assets/i18n` (only what we need for the app, though, might write a script for that or automate it)
  - `make build_locale`

arb files are nice, but the way flutter/dart_intl deals with them is pain.

### CODE REVIEW

**Comment by Komposten** [comment]
I would *highly* recommend placing all hardcoded localisation keys in a separate class:

`_localizationService.trans(Strings. authCreateAcc_previous);`
instead of
`_localizationService.trans('auth__create_acc__previous');`

This makes it a lot easier to:
- Find existing keys to use or similar ones to reuse wording. 
- Use the correct format when adding keys. 
- Prevent misspellings (when typing the key into `.trans()`.
- Edit existing keys.

**Comment by Komposten** [comment]
Since Github refuses to let me edit the above comment, I'll add some stuff here.

Using some dart magic you could even build a custom enum (which stores the string keys inside the enum constants, without requiring the constants' names to be the long underscore-y key strings) so that `trans()` only can be passed correct values.

**Comment by jozsefsallai** [comment]
@Komposten Fair enough, that's something that I would love to be able to do too, *but* the command that generates the arb files looks at the getters it can find in `lib/services/localization.dart`, and ignores pretty much everything else.

`_localizationService.auth__create_acc__previous` works too, and is intellisense-compatible. But that method decreases overall code readability.

**Comment by Komposten** [comment]
Can those getters be static? If they can a Strings class could simply be used for convenience (shorter class and key names):

```
// In a Strings class
static get authCreateAcc_previous => LocalizationService.auth__create_acc__previous
```
(Pseudocode, since I'm on my phone and not the best at dart syntax.) 

Will have better options for looking at the whole picture when I have access to my desktop.

**Comment by lifenautjoe** [comment]
I believe @uiboy merged this on his branch

**Comment by jozsefsallai** [comment]
@Komposten I get where you're coming from, but Dart uses the getters that are directly [defined in the `LocalizationService` class, inside the localization.dart file](https://github.com/jozsefsallai/openbook-app/blob/master/lib/services/localization.dart#L33), to generate the arb files. So creating new files and new classes to store the strings in would make dart fail at generating the arb files. 

Like I said, `_localizationService.auth__create_acc__previous` works perfectly and has intellisense, but referring to them like this will make the code messy and less readable.

I specifically created a new function inside the class (trans()) that returns the necessary string for any provided key (and throws an error if it can't find it (iirc)) so that the code is readable (and also compatible with how we did i18n in openbook initially). 

I preferred the json way a lot more, but the arb method is not only faster but also more feature-rich. Maybe in the future I will write a more extensible json i18n package to solve the caveats we're currently having with both the json and the arb method.

**Comment by jozsefsallai** [comment]
@lifenautjoe I forgot to mention that some of the strings (mainly the ones you have to pass variables to) need to be looked over, since I'm sure the json variable syntax is not the same as the arb one. Luckily there aren't a lot of those.

**Comment by Komposten** [comment]
@jozsefsallai 
> I get where you're coming from, but Dart uses the getters that are directly [defined in the `LocalizationService` class, inside the localization.dart file](https://github.com/jozsefsallai/openbook-app/blob/master/lib/services/localization.dart#L33), to generate the arb files. So creating new files and new classes to store the strings in would make dart fail at generating the arb files.

I do understand this. I'm asking if it is possible to make those getters _static_ members instead of instance members:
```dart
//Instance method, how it is currently
String get auth__headline {
    return Intl.message("Better social.", name: 'auth__headline');
  }

//Static method
static String get auth__headline {
    return Intl.message("Better social.", name: 'auth__headline');
  }
```

This means that the getters remain in localization.dart and can still be used to generate the .arb files (assuming that the `static` keyword doesn't break the .arb generator). However, because the getters are now static, we can simply refer to them statically from another class, and use that to get shorter/better names:
```dart
// In our new class
static get authCreateAcc_previous => LocalizationService.auth__create_acc__previous
```
-----
Alternatively, we could just skip this and go back to my original suggestion: Add a Strings class with a String (or enum) constant for every .arb key. When someone adds a getter to localization.dart they should also add the string to Strings (or we can maybe generate the Strings class from the getters in localization.dart automatically).

**Comment by lifenautjoe** [comment]
Oh oh, I’m not sure I fully understand the discussion but Shantanu already
merged this branch on his translations branch, so whatever needs to change,
nows the time CC @uiboy .

**Comment by jozsefsallai** [comment]
@Komposten Ah, I understand now. I don't think it's really worth doing that to be honest. We already have the same localization strings in:
 - localization.dart
 - arb files (generated from localization.dart)
 - dart locale files (generated from arb files)

making an extra file containing pretty much the same information/references is kind of useless and adds too much weight to the project. 

@lifenautjoe No code-related changes are required now, except for:

> I forgot to mention that some of the strings (mainly the ones you have to pass variables to) need to be looked over, since I'm sure the json variable syntax is not the same as the arb one. Luckily there aren't a lot of those.

the strings can be edited in the localization.dart file, as described in my first post

**Comment by Komposten** [comment]
@jozsefsallai It may add an additional file, but it is for convenience. Similar to how Android generates a class called `R` which contains constants for all strings, images, etc. you use in the app. Want to use that confirm string you added two days ago? No problem, just type `R.confirm` and go through the intellisense list to find it.

Makes it easier since everything is in one place (which, I guess, it technically is already in localization.dart), it has intellisense, and can be enum-ified to 100% guarantee that no one accidentally misspells anything.

It's not required by any means, and if it would end up being too much of a hassle to implement or maintain it's probably not worth it, though. 🤷‍♀️

**Comment by jozsefsallai** [comment]
@Komposten I just don't really see the point. The current getters also have good intellisense, and are referable:
image

**Comment by uiboy** [comment]
@jozsefsallai , @Komposten  Yep one of my intentions to merge everything was to look into fixing this problem.

What i did notice on the dart package `intl` documentation for .arb files (https://pub.dev/packages/intl) is that the make command with ` pub run intl_translation:extract_to_arb ` can take multiple input files instead of the one `localization.dart`. 

So Im guessing we could split the strings into their own classes that way and consolidate them in the make command. (theoretically might work, even though the flutter tutorial just uses one file )

` pub run intl_translation:extract_to_arb --output-dir=dir authLocalizations.dart settingsLocalizations.dart etc.`

However, that comes with its own challenges... one of them being we would have to now include all these various classes into the bootstrapping of `LocalizationDelegates`, where currently we only set one service (`LocalizationService`) and re-initialise them when a locale changes.  

In my branch I have been working to incorporate changing locales and loading the new translation seamlessly.. it can get tricky if we need to initialise multiple classes. 

Also then each widget that uses localization, would have to include the respective class from which it wants to use the string of, for eg `AuthLocalizationService`, `SettingsLocalizationService` etc... 

If there would be a way to consolidate all these separate classes via a layer maybe its something to look into. ( I want to try that but probably next not as part of my current PR ) 

So for now I think it does make sense to just accept that all the strings will be in `localizations.dart` with the naming scheme and continue using it as @jozsefsallai  originally intended. 

Btw just for more context,  the PR that Im creating is for translating post comments and text .. by clicking a 'see translation' button. As part of that PR, obviously there had to be a way to choose the current language of the user and change it which affects the language the post is translated into. Im also taking into account the device locale and setting it as a default.

**Comment by lifenautjoe** [comment]
Closing as was merged on Shantanu's PR!
